src/main/java/fi/aalto/cs/apluscourses/intellij/actions/ImportProjectSettingsAction.java

```diff
@@ -0,0 +1,133 @@
+package fi.aalto.cs.apluscourses.intellij.actions;
+
+import com.intellij.openapi.actionSystem.AnAction;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.project.ProjectManager;
+import com.intellij.openapi.util.io.FileUtilRt;
+import fi.aalto.cs.apluscourses.intellij.services.MainViewModelProvider;
+import fi.aalto.cs.apluscourses.intellij.services.PluginSettings;
+import fi.aalto.cs.apluscourses.model.Course;
+import fi.aalto.cs.apluscourses.model.UnexpectedResponseException;
+import fi.aalto.cs.apluscourses.presentation.CourseViewModel;
+import fi.aalto.cs.apluscourses.ui.IntelliJDialogs;
+import fi.aalto.cs.apluscourses.ui.base.Dialogs;
+import fi.aalto.cs.apluscourses.utils.CoursesClient;
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.stream.Collectors;
+import net.lingala.zip4j.ZipFile;
+import net.lingala.zip4j.model.FileHeader;
+import org.jetbrains.annotations.NotNull;
+
+public class ImportProjectSettingsAction extends AnAction {
+  @NotNull
+  private MainViewModelProvider mainViewModelProvider;
+
+  @NotNull ProjectSettingsImporter projectSettingsImporter;
+
+  @NotNull
+  private Dialogs dialogs;
+
+  /**
+   * Construct an action with the given main view model provider, project settings importer and
+   * dialog helper.
+   */
+  public ImportProjectSettingsAction(@NotNull MainViewModelProvider mainViewModelProvider,
+                                     @NotNull ProjectSettingsImporter projectSettingsImporter,
+                                     @NotNull Dialogs dialogs) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.projectSettingsImporter = projectSettingsImporter;
+    this.dialogs = dialogs;
+  }
+
+  public ImportProjectSettingsAction() {
+    this(PluginSettings.getInstance(), ImportProjectSettingsAction::doImport,
+        new IntelliJDialogs());
+  }
+
+  @Override
+  public void actionPerformed(@NotNull AnActionEvent e) {
+    Project project = e.getProject();
+    if (project == null) {
+      informNoProjectOpen();
+      return;
+    }
+
+    CourseViewModel courseViewModel = mainViewModelProvider
+        .getMainViewModel(project)
+        .courseViewModel
+        .get();
+    if (courseViewModel == null) {
+      return;
+    }
+    Course course = courseViewModel.getModel();
+
+    URL settingsUrl = course.getResourceUrls().get("projectSettings");
+    if (settingsUrl == null) {
+      informNoSettings(course);
+      return;
+    }
+
+    try {
+      projectSettingsImporter.doImport(project, settingsUrl);
+    } catch (IOException | UnexpectedResponseException ex) {
+      informErrorOccurred();
```

Comment:
Nit: we probably need to come our with a common approach for exception handling (especially taking into account recent discoveries when debugging the IDE) apart from users notifications

Reply:
Though I agree with you on that, this implementation seems mostly appropriate to me. It is in a method that is called by the platform, so it cannot be thrown any further. Of course, some information of the exception could be used to tailor a more informative/specific error message and stack trace could be printed to the log. I also like `Dialogs` interface and `showErrorDialog()` method there. Perhaps that approach could be abstracted even further so that we had some kind of `ErrorHandler` interface, an implementation of which would handle both logging and informing user.

Reply:
the implementation is good, indeed, and this idea would probably make it perfect:
> Perhaps that approach could be abstracted even further so that we had some kind of `ErrorHandler` interface, an implementation of which would handle both logging and informing user.